File: codejam/python/2009/Qual/Qual_Watersheds.py
import sys
from collections import deque

## Input crap
infile = sys.stdin
intokens = deque()
def getTokens(): 
    while not intokens:
        for c in infile.readline().rstrip().split() :
            intokens.append(c)    
def gs(): getTokens(); return intokens.popleft()
def gi(): return int(gs())
def gf(): return float(gs())
def gbs(): return [c for c in gs()]
def gis(n): return [gi() for i in range(n)]
def ia(m): return [0] * m
def iai(m,v): return [v] * m
def twodi(n,m,v): return [iai(m,v) for i in range(n)]
def fill2(m) : r = gis(2*m); return r[0::2],r[1::2]
def fill3(m) : r = gis(3*m); return r[0::3],r[1::3],r[2::3]
def fill4(m) : r = gis(4*m); return r[0::4],r[1::4],r[2::4],r[3::4]
MOD = 998244353
##MOD = 1000000007

class dsu :
    def __init__(self,n=1) :
        self.n = n
        self.parentOrSize = [-1 for i in range(n)]
    def leader(self,a) :
        if self.parentOrSize[a] < 0 : return a
        ans = self.leader(self.parentOrSize[a])
        self.parentOrSize[a] = ans
        return ans
    def merge(self,a,b) :
        x = self.leader(a); y = self.leader(b)
        if x == y : return x
        if self.parentOrSize[y] < self.parentOrSize[x] : (x,y) = (y,x)
        self.parentOrSize[x] += self.parentOrSize[y]
        self.parentOrSize[y] = x
        return x
    def same(self,a,b) :
        return self.leader(a) == self.leader(b)
    def size(self,a) :
        l = self.leader(a)
        return -self.parentOrSize[l]
    def groups(self) :
        leaderBuf = [0 for i in range(self.n)]
        groupSize = [0 for i in range(self.n)]
        for i in range(self.n) :
            leaderBuf[i] = self.leader(i)
            groupSize[leaderBuf[i]] += 1
        preres = [ [] for i in range(self.n) ]
        for (i,v) in enumerate(leaderBuf) :
            preres[v].append(i)
        return [x for x in preres if x]

def main() :
    if len(sys.argv) > 1 : global infile; infile = open(sys.argv[1],'rt')
    ## PROGRAM STARTS HERE
    T = gi()
    for tt in range(1,T+1) :
        H,W = gi(),gi(); gr = [ gis(W) for _ in range(H) ]
        du = dsu(H*W)
        for i in range(H) :
            for j in range(W) :
                best = gr[i][j]; ii,jj = -1,-1
                if i-1 >= 0 and gr[i-1][j] < best : best,ii,jj = gr[i-1][j],i-1,j
                if j-1 >= 0 and gr[i][j-1] < best : best,ii,jj = gr[i][j-1],i,j-1
                if j+1 < W  and gr[i][j+1] < best : best,ii,jj = gr[i][j+1],i,j+1
                if i+1 < H  and gr[i+1][j] < best : best,ii,jj = gr[i+1][j],i+1,j
                if ii != -1 : du.merge(W*ii+jj,W*i+j)
        print(f"Case #{tt}:")
        lookup = [-1] * (H*W)
        alph = "abcdefghijklmnopqrstuvwxyz"
        curs = 0
        for i in range(H*W) :
            if lookup[du.leader(i)] == -1 : lookup[du.leader(i)] = alph[curs]; curs += 1
        for i in range(H) :
            line = " ".join([lookup[du.leader(W*i+j)] for j in range(W)])
            print(line)

if __name__ == "__main__" :
    main()

File: codejam/cpp/2009/Qual/Qual_Watersheds.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<ll> vi;
typedef pair<ll,ll> pi;
#define FOR(i,a) for (ll i = 0; i < (a); i++)
#define len(x) (ll) x.size()
const ll INF = 1LL << 62;
const ll MOD = 1000000007;
//const ll MOD = 998244353;
const double PI = 4*atan(double(1.0));

struct dsu {
    private :
        ll _n;
        vi parentOrSize;
    public :
        dsu() : _n(0) {}
        dsu(ll n) : _n(n),parentOrSize(n,-1) {}
        ll leader(ll a) {
            ll x = parentOrSize[a];
            if (x < 0) { return a; }
            return parentOrSize[a] = leader(x);
        }
        ll merge(ll a, ll b) {
            ll x = leader(a); ll y = leader(b);
            if (x == y) { return x; }
            if (parentOrSize[y] < parentOrSize[x]) { swap(x,y); }
            parentOrSize[x] += parentOrSize[y];
            return parentOrSize[y] = x;
        }
        bool same(ll a, ll b) {
            return leader(a) == leader(b);
        }
        ll size(ll a) {
            auto x = leader(a);
            return -parentOrSize[x];
        }
        vector<vector<ll>> groups() {
            vi leaderBuf(_n,0), leaders;
            FOR(i,_n) { if (i == leader(i)) { leaders.push_back(i);} }
            ll nlead = len(leaders);
            vi lkup(_n,-1); FOR(i,nlead) { lkup[leaders[i]]=i; }
            vector<vector<ll>> ans(nlead);
            FOR(i,_n) { ans[lkup[leader(i)]].push_back(i); }
            return ans;
        }
};

int main (int argc, char **argv) {
    if (argc > 1) { freopen(argv[1],"r",stdin); }
    ios_base::sync_with_stdio(false);cin.tie(0);
    // PROGRAM STARTS HERE
    ll gr[110][110];
    ll T; cin >> T;
    for (ll tt=1;tt<=T;tt++) {
        ll H,W; cin >> H >> W;
        FOR(i,H) FOR(j,W) cin >> gr[i][j];
        dsu du(H*W);
        FOR(i,H) {
            FOR(j,W) {
                ll best = gr[i][j]; ll ii = -1; ll jj = -1;
                if (i-1 >= 0 && gr[i-1][j] < best) { best = gr[i-1][j]; ii = i-1; jj = j; }
                if (j-1 >= 0 && gr[i][j-1] < best) { best = gr[i][j-1]; ii = i; jj = j-1; }
                if (j+1  < W && gr[i][j+1] < best) { best = gr[i][j+1]; ii = i; jj = j+1; }
                if (i+1  < H && gr[i+1][j] < best) { best = gr[i+1][j]; ii = i+1; jj = j; }
                if (ii != -1) { du.merge(W*ii+jj,W*i+j); }
            }
        }
        cout << "Case #" << tt << ":\n";
        vector<char> lookup(H*W,'.'); char c = 'a';
        FOR(i,H*W) {
            auto l = du.leader(i);
            if (lookup[l] == '.') { lookup[l] = c; c++; }
            cout << lookup[l] << ((i%W==W-1) ? '\n' : ' ');
        }
    }
}

